### PR TITLE
Add command manager

### LINKED ISSUE #24: Switch to command manager
I opted out of this feature because it's pretty complex, and I figured we'd only have 2 or 3 commands, but since it's racking up new ones really quickly, we might wanna go for something that manages these commands. It means that we could output help for all of those as well. 

Whoever takes this on, there's a command manager here for our language here: https://github.com/discordjs/Commando

### PR DESCRIPTION
**Pull request by stelar7**
fixes #24 (except it doesnt use that lib...)

### CODE REVIEW

File: settings/shared_settings.json [hunk 1]
```diff
@@ -45,24 +55,37 @@
...
-        "baseURL":"https://api.github.com/repos/WxWatch/riot-api-libraries/contents/libraries/"
+        "baseURL": "https://api.github.com/repos/WxWatch/riot-api-libraries/contents/libraries/"
     },
     "autoReact": {
-        "emoji": "408527155891273738"
+        "emoji": "408527155891273738",
+        "aliases":[
+            "toggle_default_thinking",
+            "toggle_react",
+            "refresh_thinking"
```
Comment (Querijn) [inline settings/shared_settings.json]: not very descriptive

also it stops reacting to hi

File: src/OfficeHours.ts [hunk 2]
```diff
@@ -79,63 +79,62 @@ export default class OfficeHours {
         }
     }
 
-    onCommand(message: Discord.Message) {
+    onCommand(sender: Discord.User, channel: Discord.TextChannel, message: Discord.Message, command: string, args: string[]) {
```
Comment (Querijn) [inline src/OfficeHours.ts]: sender and channel are in message, why are they separated?

File: src/Botty.ts [hunk 3]
```diff
@@ -88,4 +89,32 @@ export default class Botty {
     start() {
         return this.client.login(this.personalSettings.discord.key);
     }
+
+    private commands: CommandBase[] = [];
```
Comment (molenzwiebel) [inline src/Botty.ts]: Move this up with the other instance variables.

File: src/Botty.ts [hunk 4]
```diff
@@ -88,4 +89,32 @@ export default class Botty {
...
+
+        if (message.author.bot) return;
+        if (!message.content.startsWith(this.sharedSettings.botty.prefix)) return;
+
+        const parts = message.content.split(" ");
+        const command = parts[0].substr(1);
+
+        this.commands.forEach(cmd => {
+            if (cmd.alias === command) {
+                cmd.handler.onCommand(message.author, <Discord.TextChannel>message.channel, message, command, parts.slice(1))
```
Comment (molenzwiebel) [inline src/Botty.ts]: Missing `;`

File: src/OfficeHours.ts [hunk 5]
```diff
@@ -19,29 +20,28 @@ interface OfficeHoursData {
...
-        this.bot.on("ready", this.onBot.bind(this));
-        this.bot.on("message", this.onCommand.bind(this));
-		// this.bot.on("channelUpdate", this.onChannelUpdate.bind(this));
+        // this.bot.on("channelUpdate", this.onChannelUpdate.bind(this));
     }
 
-    onBot() {
-        const mainGuild = this.bot.guilds.get(this.sharedSettings.server);
+    onReady(bot: Discord.Client) {
+        this.bot = bot;
```
Comment (molenzwiebel) [inline src/OfficeHours.ts]: Any cleaner way to do this? This seems a bit hacky.

File: src/OfficeHours.ts [hunk 6]
```diff
@@ -151,28 +150,28 @@ export default class OfficeHours {
...
-        .then(event => { 
-            channel.send(`On this day in ${event.year}, ${event.text}`);
-        })
-        .catch(r => console.warn('Failed to fetch message of the day: ' + r));
+            .then(r => r.json())
+            .then(({ data: { Events } }) => Events[Math.floor(Math.random() * Events.length)])
+            .then(event => {
+                channel.send(`On this day in ${event.year}, ${event.text}`);
+            })
+            .catch(r => console.warn('Failed to fetch message of the day: ' + r));
```
Comment (molenzwiebel) [inline src/OfficeHours.ts]: Might be cleaner to convert this to an `async` method now that we're here anyway 😉

File: src/OfficeHours.ts [hunk 7]
```diff
@@ -192,40 +191,40 @@ export default class OfficeHours {
...
+                    if (usersToMention.length > 0)
+                        channel.send(usersToMention.join(", ") + "\n");
+                }
+
+                this.sendOnThisDayMessage(channel);
+            })
+            .catch(reason => {
+                console.warn("Failed getting last close message: " + reason);
+                this.sendOnThisDayMessage(channel);
+            });
```
Comment (molenzwiebel) [inline src/OfficeHours.ts]: This can also be async (the method is already async lul)

File: src/RiotAPILibraries.ts [hunk 8]
```diff
@@ -71,58 +67,31 @@ export default class RiotAPILibraries {
...
-            }
+        return this.getListForLanguage(channel, param);
+    }
 
-            let editMessage = await editMessagePromise; 
-            if (Array.isArray(editMessage)) editMessage = editMessage[0];
-            editMessage.edit({ embed });
-        }
+    invalidArguments(channel: Discord.TextChannel, arg: string[]) {
+        channel.send(`unknown argument for command; ${arg}`);
```
Comment (molenzwiebel) [inline src/RiotAPILibraries.ts]: This can be a lot cleaner than this ;)

File: src/RiotAPILibraries.ts [hunk 9]
```diff
@@ -150,6 +119,47 @@ export default class RiotAPILibraries {
...
+            return;
+        }
+
+        const libraryList = await response.json();
+        if (!Array.isArray(libraryList) || libraryList.length === 0 || !libraryList[0].sha) {
+            channel.send(this.settings.riotApiLibraries.noLanguage + language);
+            return;
+        }
+
+        let editMessagePromise = channel.send(`Found the list of libraries for ${language}, listing ${libraryList.length} libraries, this post will be edited with the result.`);
```
Comment (molenzwiebel) [inline src/RiotAPILibraries.ts]: `const`

File: settings/shared_settings.json (hunk 1 above)
Comment (stelar7) [reply settings/shared_settings.json]: I haven't touched the react code, so that should still work.. :thinking:

File: src/OfficeHours.ts (hunk 5 above)
Comment (stelar7) [reply src/OfficeHours.ts]: Its either gonna end up in onReady or the constructor, and I'd prefer having the constructors as clean as possible :man_shrugging: 

(Tho I managed to remove it altogether from officehours, it still remains in some of the other classes)

File: settings/shared_settings.json (hunk 1 above)
Comment (Querijn) [reply settings/shared_settings.json]: No what I mean is that the "thinking" isn't the only reaction it has, it also stops the triggers for greetings

**Comment by Querijn** [comment]
Gonna put this one on hold. The issue I currently have is that the command's intention is completely hidden.

Can you tell what this function does?
afbeelding

Usually the startsWith("!libs") gave it away immediately, but since that line is now gone, it's hard to tell which is which, especially since the calls in app.ts don't give it away either. It took me a while to find out which one actually was used for help.

I have this setup in the "stelar7-commands" branch, which I'm going to keep out of main branch for two reasons, the first being that !help is currently in use by Heimerdinger, which will go down as soon as it's merged with Botty, and the second being issues described above. I'll merge it if I find a good way to more easily define which function does what exactly.

**Pull request closed without merging**